pkg/controller/service_lb.go

```diff
@@ -0,0 +1,392 @@
+		if err != nil {
+			klog.Errorf("failed to update deployment %s, err: %v", deploy.Name, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) getLbSvcPod(svcName, svcNamespace string) (*corev1.Pod, error) {
+	sel, _ := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{"app": genLbSvcDpName(svcName), "namespace": svcNamespace},
+	})
+
+	pods, err := c.podsLister.Pods(svcNamespace).List(sel)
+	if err != nil {
+		return nil, err
+	} else if len(pods) == 0 {
+		time.Sleep(2 * time.Second)
+		return nil, fmt.Errorf("pod '%s' not exist", genLbSvcDpName(svcName))
+	} else if len(pods) != 1 {
+		time.Sleep(2 * time.Second)
+		return nil, fmt.Errorf("too many pod")
+	} else if pods[0].Status.Phase != "Running" {
+		time.Sleep(2 * time.Second)
+		return nil, fmt.Errorf("pod is not active now")
+	}
+
+	return pods[0], nil
+}
+
+func (c *Controller) validateSvc(svc *corev1.Service) error {
+	providerName := getAttachNetworkProvider(svc)
+	attchSubnetAnnotation := fmt.Sprintf(util.LogicalSwitchAnnotationTemplate, providerName)
+
+	if svc.Spec.LoadBalancerIP != "" {
+		if ip := net.ParseIP(svc.Spec.LoadBalancerIP); ip == nil {
+			return fmt.Errorf("Invalid static loadbalancerIP %s for svc %s", svc.Spec.LoadBalancerIP, svc.Name)
+		}
+	}
+
+	if svc.Annotations[attchSubnetAnnotation] != "" {
+		subnet, err := c.subnetsLister.Get(svc.Annotations[attchSubnetAnnotation])
+		if err != nil {
+			klog.Errorf("failed to get subnet %v", err)
+			return err
+		}
+
+		if svc.Spec.LoadBalancerIP != "" && !util.CIDRContainIP(subnet.Spec.CIDRBlock, svc.Spec.LoadBalancerIP) {
+			return fmt.Errorf("The loadbalancerIP %s is not in the range of subnet %s, cidr %v", svc.Spec.LoadBalancerIP, subnet.Name, subnet.Spec.CIDRBlock)
+		}
+	}
+	return nil
+}
+
+func (c *Controller) getPodAttachIP(pod *corev1.Pod, svc *corev1.Service) (string, error) {
+	var loadBalancerIP string
+	var err error
+
+	providerName := getAttachNetworkProvider(svc)
+	attchIpAnnotation := fmt.Sprintf(util.IpAddressAnnotationTemplate, providerName)
+
+	if pod.Annotations[attchIpAnnotation] != "" {
+		loadBalancerIP = pod.Annotations[attchIpAnnotation]
+	} else {
+		err = fmt.Errorf("failed to get attachment ip from pod's annotation")
+	}
+
+	return loadBalancerIP, err
+}
+
+func (c *Controller) deleteLbSvc(svc *corev1.Service) error {
+	if _, err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).Get(context.Background(), genLbSvcDpName(svc.Name), metav1.GetOptions{}); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	if err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).Delete(context.Background(), genLbSvcDpName(svc.Name), metav1.DeleteOptions{}); err != nil {
+		klog.Errorf("failed to delete deployment %s, err: %v", genLbSvcDpName(svc.Name), err)
+		return err
+	}
+
+	for {
```

Comment:
no need to wait pod deletion

Reply:
Modified.
This is used when the net-attach-def is created for each pod, and delete net-attach-def after pod is deleted.